File: app/src/main/java/com/yatatsu/passiveviewsample/ui/repos/ReposController.java
package com.yatatsu.passiveviewsample.ui.repos;


import com.yatatsu.passiveviewsample.data.api.GitHubService;
import com.yatatsu.passiveviewsample.data.model.Repository;
import com.yatatsu.passiveviewsample.ui.base.Controller;

import java.util.List;

import rx.functions.Action1;

public class ReposController extends Controller<ReposScreen> {

    private final GitHubService github;
    private String username;

    public ReposController(GitHubService github) {
        this.github = github;
    }

    void onCreateScreen(String username) {
        this.username = username;
    }

    void onStartScreen() {
        register(github.getRepositories(username)
                .subscribe(new Action1<List<Repository>>() {
                    @Override
                    public void call(List<Repository> repositories) {
                        getScreen().showRepositories(repositories);
                    }
                }, new Action1<Throwable>() {
                    @Override
                    public void call(Throwable throwable) {
                        throwable.printStackTrace();
                        getScreen().showError(throwable.getMessage());
                    }
                }));
    }

    void onStopScreen() {
        unSubscribe();
    }
}

File: app/src/main/java/com/yatatsu/passiveviewsample/ui/repos/ReposController.kt
package com.yatatsu.passiveviewsample.ui.repos


import com.yatatsu.passiveviewsample.data.api.GitHubService
import com.yatatsu.passiveviewsample.data.model.Repository
import com.yatatsu.passiveviewsample.ui.ApiErrorPresenter
import com.yatatsu.passiveviewsample.ui.base.Controller

class ReposController(private val github: GitHubService) : Controller<ReposScreen>() {
  private var username: String? = null

  fun onCreateScreen(username: String?) {
    this.username = username
  }

  fun onStartScreen() {
    register(github.getRepositories(username!!).subscribe(
        { repositories -> screen?.showRepositories(repositories) }) { throwable ->
      throwable.printStackTrace()
      screen?.showError(
          ApiErrorPresenter.getErrorMessage(throwable))
    })
  }

  fun onStopScreen() {
    unSubscribe()
  }

  fun onItemClick(repository: Repository) {
    screen?.openUrl(repository.htmlUrl!!)
  }
}
